Userland/Applications/PixelPaint/Tools/TextTool.cpp

```diff
@@ -0,0 +1,286 @@
+            }
+            auto line_selection_length = end_col - start_col;
+            auto selected_string = m_text_editor->line(i).view().substring_view(start_col, line_selection_length);
+            auto text_before_selection = m_text_editor->line(i).view().substring_view(0, start_col);
+            auto selected_width = m_selected_font->width(selected_string);
+            auto selection_x_offset = m_selected_font->width(text_before_selection);
+
+            // the + 4 here is because that's how Painter::do_draw_text calculates line height, instead of asking
+            // the font it's preferred line height. If we don't replicate that here, the letters jump around when they
+            // get selected.
+            auto selection_y_offset = static_cast<int>((m_selected_font->pixel_size() + 4) * i);
+
+            auto selection_rect = Gfx::IntRect(selection_x_offset, selection_y_offset, selected_width, m_selected_font->preferred_line_height());
+            text_painter.fill_rect(selection_rect, m_text_editor->palette().selection());
+            text_painter.draw_text(selection_rect, selected_string, Gfx::TextAlignment::TopLeft, m_text_editor->palette().selection_text());
+        }
+    }
+
+    /* auto scaled_bitmap = text_bitmap->scaled(m_editor->scale(), m_editor->scale()).release_value();
+    painter.blit({ text_location.x(), text_location.y() }, scaled_bitmap, scaled_bitmap->rect()); */
+    auto scaled_width = static_cast<int>(ceilf(m_editor->scale() * static_cast<float>(text_bitmap->width())));
+    auto scaled_height = static_cast<int>(ceilf(m_editor->scale() * static_cast<float>(text_bitmap->height())));
+    auto scaled_rect = Gfx::IntRect(text_location.x(), text_location.y(), scaled_width, scaled_height);
+    scaled_rect.set_location({ text_location.x(), text_location.y() });
+    painter.draw_scaled_bitmap(scaled_rect, text_bitmap, text_bitmap->rect(), 1.0);
+
+    // marching ants box
+    auto right_padding = m_selected_font->width("  "sv);
+    m_ants_rect = Gfx::IntRect(text_location.translated(-4, -2), { scaled_rect.width() + 4 + right_padding, scaled_rect.height() + 4 });
+    m_editor->draw_marching_ants(painter, m_ants_rect);
+
+    // blinking cursor
+    if (m_cursor_state) {
+        auto cursor_text_prefix = m_text_editor->line(m_text_editor->cursor().line()).to_utf8().substring(0, m_text_editor->cursor().column());
+        auto cursor_x_offset = static_cast<int>((m_selected_font->width(cursor_text_prefix) + 1) * m_editor->scale());
+        auto cursor_y_offset = static_cast<int>((m_text_editor->cursor().line() * (m_selected_font->pixel_size() + 4)) * m_editor->scale());
+
+        auto cursor_width = max<int>(1 * m_editor->scale(), 1);
+        auto cursor_height = (m_selected_font->pixel_size() + 4) * m_editor->scale();
+        auto cursor_rect = Gfx::Rect<int>(text_location.x() + cursor_x_offset, text_location.y() + cursor_y_offset, cursor_width, cursor_height);
```

Comment:
This is a heck of a lot of work to get a cursor rect. Is there any way we can leverage the work already done by `m_text_editor` to simplify this?

Reply:
This was actually pretty easy! Thanks for having me take another look at it. The newest code lets TextEditor say where the cursor is, we then scale it based on ImageEditor scale() and then position it where the text is on screen and draw. About as many lines due needing to scale an IntRect by a float, but certainly easier to follow than before.